internal/ledger/compare_test.go

```diff
@@ -0,0 +1,599 @@
+		},
+	}
+
+	// Run test cases individually
+	for testName, testCase := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			// Create temporary directories for the sample snapshots and comparison results
+			snapshotDir1, err := ioutil.TempDir("", "sample-snapshot-dir1")
+			require.NoError(t, err)
+			defer os.RemoveAll(snapshotDir1)
+			snapshotDir2, err := ioutil.TempDir("", "sample-snapshot-dir2")
+			require.NoError(t, err)
+			defer os.RemoveAll(snapshotDir2)
+			resultsDir, err := ioutil.TempDir("", "results")
+			require.NoError(t, err)
+			defer os.RemoveAll(resultsDir)
+
+			// Populate temporary directories with sample snapshot data
+			err = createSnapshot(snapshotDir1, testCase.inputTestRecords1, testCase.inputSignableMetadata1)
+			require.NoError(t, err)
+			err = createSnapshot(snapshotDir2, testCase.inputTestRecords2, testCase.inputSignableMetadata2)
+			require.NoError(t, err)
+
+			// Compare snapshots and check the output
+			out, err := compareSnapshots(snapshotDir1, snapshotDir2, filepath.Join(resultsDir, "results.json"))
+			switch testCase.expectedOutputType {
+			case "error":
+				require.ErrorContains(t, err, testCase.expectedOutput)
+			case "json":
+				require.NoError(t, err)
+				require.JSONEq(t, testCase.expectedOutput, out)
+			default:
+				panic("unexpected code path: bug")
+			}
+		})
+	}
+}
+
+// createSnapshot generates a sample snapshot based on the passed in records and metadata
+func createSnapshot(dir string, pubStateRecords []*testRecord, signableMetadata *kvledger.SnapshotSignableMetadata) error {
+	// Generate public state of sample snapshot
+	pubStateWriter, err := privacyenabledstate.NewSnapshotWriter(
+		dir,
+		privacyenabledstate.PubStateDataFileName,
+		privacyenabledstate.PubStateMetadataFileName,
+		testNewHashFunc,
+	)
+	if err != nil {
+		return err
+	}
+	defer pubStateWriter.Close()
+
+	// Add sample records to sample snapshot
+	for _, sample := range pubStateRecords {
+		err = pubStateWriter.AddData(sample.namespace, &privacyenabledstate.SnapshotRecord{
+			Key:      []byte(sample.key),
+			Value:    []byte(sample.value),
+			Metadata: []byte(sample.metadata),
+			Version:  toBytes(sample.blockNum, sample.txNum),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	_, _, err = pubStateWriter.Done()
+	if err != nil {
+		return err
+	}
+
+	// Generate the signable metadata files for sample snapshot
+	signableMetadataBytes, err := signableMetadata.ToJSON()
+	if err != nil {
+		return err
+	}
+	// Populate temporary directory with signable metadata files
+	err = fileutil.CreateAndSyncFile(filepath.Join(dir, kvledger.SnapshotSignableMetadataFileName), signableMetadataBytes, 0o444)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// compareSnapshots calls the Compare tool and extracts the result json
+func compareSnapshots(ss1 string, ss2 string, res string) (string, error) {
+	// Run compare tool on snapshots
+	_, err := Compare(ss1, ss2, res)
```

Comment:
This function now returns `counts` as well. Please enhance the test to verify the counts as well.